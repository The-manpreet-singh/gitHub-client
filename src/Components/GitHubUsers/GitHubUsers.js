import "./GitHubUsers.scss";
import { useState, useEffect } from "react";
import axios from "axios";
import { Link } from "react-router-dom";
import {
  Container,
  Grid,
  Avatar,
  Card,
  CardContent,
  Typography,
  Button,
} from "@mui/material";
import InfiniteScroll from "react-infinite-scroll-component";

export default function GitHubUsers() {
  const [page, setPage] = useState(1);
  const [users, setUsers] = useState([]);
  const [sorting, setSorting] = useState([]);
  const [sortType, setSortType] = useState("");

  useEffect(() => {
    const loadUsers = async () => {
      // const res =
      //   await axios.get(`https://api.github.com/search/users?q=language:javascript+type:user&sort=followers&order
      // =desc`);
      // const data = [...res.data.items];

      // const githubToken = "713d68d65422abe4eed50fbf9cd000ddfbbd28fa";
      // const res = await axios({
      //   method: "get",
      //   url: `https://api.github.com/users`,
      //   headers: {
      //     Authorization: `token ${githubToken}`,
      //     "Content-Type": "application/json",
      //     "Access-Control-Allow-Origin": "*",
      //     "Access-Control-Allow-Credentials": "true",
      //     "Access-Control-Allow-Methods": "GET,HEAD,OPTIONS,POST,PUT",
      //     "Access-Control-Allow-Headers":
      //       "Origin, X-Requested-With, Content-Type, Accept, Authorization",
      //   },
      // });
      // const data = [...res.data];

      const res = await axios.get(`https://api.github.com/users`);
      const data = [...res.data];
      //console.log(data);

      var userUrl = [];
      for (let i = 0; i < data.length; i++) {
        userUrl.push(data[i].login);

        //console.log(userUrl);
      }

      const res1 = [];
      for (let i = 0; i < userUrl.length; i++) {
        const res2 = await axios.get(
          ` https://api.github.com/users/${userUrl[i]}`
        );

        res1.push(res2.data);

        //console.log(res2.data);
      }
      //console.log(res1);
      setUsers(res1);
    };
    loadUsers();
  }, []);

  useEffect(() => {
    const sortArray = (type) => {
      const types = {
        followers: "followers",
        public_repos: "public_repos",
      };
      const sortProperty = types[type];
      const sorted = [...users].sort(
        (a, b) => b[sortProperty] - a[sortProperty]
      );
      //  console.log(sorted);
      setSorting(sorted);
    };
    sortArray(sortType);
  }, [sortType, users]);

  const gitHubList = (
    <InfiniteScroll
      dataLength={sorting.length}
      next={() => setPage(page + 1)}
      hasMore={true}
    >
      <Grid container spacing={2} className="dataStyle">
        {sorting.map((user) => (
          <Grid item sm={3} key={user.id}>
            <Card>
              <CardContent>
                <Typography component={"div"} className="Avatar">
                  <Avatar
                    alt="Remy Sharp"
                    src={user.avatar_url}
                    sx={{ width: 200, height: 200 }}
                  />
                </Typography>
                <Typography className="Card" component={"div"} variant="body2">
                  Name: {user.name}
                  <br />
                  Login Name: {user.login} <br />
                  Location: {user.location} <br />
                  Repos: {user.public_repos} <br />
                  Gists: {user.public_gists} <br />
                  Followers: {user.followers} <br />
                  Following:{user.following} <br />
                </Typography>
                <Typography className="Card" component={"div"}>
                  <Button variant="outlined" size="medium">
                    <Link
                      to={{
                        pathname: `/repos/${user.login}`,
                        state: { user: user.login },
                      }}
                    >
                      Repos List
                    </Link>
                  </Button>
                </Typography>
              </CardContent>
            </Card>
          </Grid>
        ))}
      </Grid>
    </InfiniteScroll>
  );

  return (
    <div data-testid="gitHub-1" className="User">
      <Container>
        <select
          defaultValue="Sort"
          onChange={(e) => setSortType(e.target.value)}
        >
          <option disabled value="Sort">
            Sort By
          </option>
          <option value="followers">Followers</option>
          <option value="public_repos">Repos</option>
        </select>

        {gitHubList}

        {/* <Pagination
          className="Pagination"
          count={10}
          variant="outlined"
          shape="rounded"
          page={page}
          onChange={(event, value) => setPage(value)}
        />
         */}
      </Container>
    </div>
  );
}
