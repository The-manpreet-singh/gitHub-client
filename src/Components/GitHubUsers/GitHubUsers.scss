$margin-user: 20px;
$margin-pagination: 50px 0px;
$display: flex;
$justify-content: center;
$padding-card: 5px;

.User {
  margin: $margin-user;
}
.dataStyle{
  justify-content: center;
}
/* .Data {
  padding: 5px;
  margin: 10px;
} */
.Pagination {
  margin: $margin-pagination;
}
.Card {
  display: $display;
  justify-content: $justify-content;
  padding: $padding-card;
}
.Avatar {
  justify-content: $justify-content;
  display: $display;
}
