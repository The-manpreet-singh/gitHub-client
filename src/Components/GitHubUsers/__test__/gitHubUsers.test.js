import React from "react";
import ReactDom from "react-dom";
import GitHubUsers from "./../GitHubUsers";
import renderer from "react-test-renderer";
import { render, screen, cleanup } from "@testing-library/react";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDom.render(<GitHubUsers />, div);
});

afterEach(() => {
  cleanup();
});

test("should render complted GitHubUsers", () => {
  render(<GitHubUsers />);
  const githubList = screen.getByTestId("gitHub-1");
  expect(githubList).toBeInTheDocument();
  expect(githubList).toHaveTextContent("Sort ByFollowersRepos");
  expect(githubList).not.toContainHTML("<select>");
});

test("matches snapshot", () => {
  const tree = renderer.create(<GitHubUsers />).toJSON();
  //console.log(tree);
  expect(tree).toMatchSnapshot();
});
