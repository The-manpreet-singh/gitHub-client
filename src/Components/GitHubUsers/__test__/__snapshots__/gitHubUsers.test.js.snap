// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="User"
  data-testid="gitHub-1"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <select
      defaultValue="Sort"
      onChange={[Function]}
    >
      <option
        disabled={true}
        value="Sort"
      >
        Sort By
      </option>
      <option
        value="followers"
      >
        Followers
      </option>
      <option
        value="public_repos"
      >
        Repos
      </option>
    </select>
    <div
      className="infinite-scroll-component__outerdiv"
      style={Object {}}
    >
      <div
        className="infinite-scroll-component "
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "height": "auto",
            "overflow": "auto",
          }
        }
      >
        <div
          className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 dataStyle css-mhc70k-MuiGrid-root"
        />
      </div>
    </div>
  </div>
</div>
`;
